Update of CP2K and its dependencies

mtaillefumier opened the pull request:
This PR contains various updates in cp2k dependencies. 

- CP2K
    - tblite library
    - SIRIUS modules
         - Added dftd4 and dftd3 support
         - Added nlcglib support
         - Added variable cell optimization
    - Added Mi300 support 
    - Added the possibility to compile dbcsr cpu only even when `+cuda` or `+rocm` are specified
    - Updated to the latest version of cp2k 2025.2
- multicharge
    - Add toml-f library
    - Update mctc-lib version
    - Update multicharge version
    - Add jonquil dependency
    - Add missing depends_on("fortran") in multicharge
- tblite: Added cmake build system
- DBCSR
    - Added Mi300 support 
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

Commit:
Add tblite dependency

- Add the Mi300 target to cp2k and dbcsr
- Add tblite to cp2k recipe
- tblite can be build with either cmake or meson

Signed-off-by: Mathieu Taillefumier <mathieu.taillefumier@free.fr>

Commit:
Bump mctc_lib versions

Signed-off-by: Mathieu Taillefumier <mathieu.taillefumier@free.fr>

Commit:
Add missing toml-f and jonquil dependencies

- Update multicharge version
- Add missing dependencies for mctc-lib

MBlaschek committed:
odc: add version 1.6.1 (#287)

new version of odc 1.6.1

Co-authored-by: Richarda Butler <39577672+RikkiButler20@users.noreply.github.com>

adamjstewart committed:
py-scipy: add v1.16.0 (#349)

JohnGouwar committed:
New packages: py-tree-sitter and py-tree-sitter-c (#336)

balos1 committed:
deprecate sundials versions 2.x.x, 3.x.x, and 4.x.x (#267)

* deprecate sundials versions 2.x.x, 3.x.x, and 4.x.x
* remove blas depends_on

yizeyi18 committed:
Add missing language dependency for multicharge, maeparser, coordgen (#155)

downloadico committed:
harminv: made harminv depend on fortran at build time. (#375)

downloadico committed:
meep: make meep depend on Fortran (#387)

MBlaschek committed:
Eccodes: add 2.40.0 & 2.41.0 (#288)

Added 2.40.0 and 2.41.0 .

Co-authored-by: Richarda Butler <39577672+RikkiButler20@users.noreply.github.com>

RobertMaaskant committed:
py-setuptools: add v80.9.0 (#260)

RobertMaaskant committed:
yarn: add v4.9.2 (#256)

RobertMaaskant committed:
go: add v1.23.10 and v1.24.4 (#254)

RobertMaaskant committed:
glab: add v1.59.2 (#252)

RobertMaaskant committed:
gcc: add c17 and c23 support (#240)

RobertMaaskant committed:
jq: set cflags when required (#239)

kwryankrattiger committed:
Move spack gpg home for darwin runners (#359)

* Move spack gpg home for darwin runners

* Kick a package

awvwgk commented on repos/spack_repo/builtin/packages/mctc_lib/package.py:
```diff
@@ -23,6 +23,27 @@ class MctcLib(MesonPackage, CMakePackage):
+
+    version("0.4.2", sha256="ce1e962c79d871d3705be590aef44f07ca296843b85e164307290f8324769406")
+    version("0.4.1", sha256="57fe4610c4fa21d0b797f88b68481c7be1e7d291daa12063caed51bee779b88a")
+    version(
+        "0.4.0-intel-2021-ubuntu-latest",
+        sha256="b838a983137db1c223359e9b9d68cf04533b82f0e7390d485713676700c07cae",
+    )
+    version(
+        "0.4.0-gnu-12-ubuntu-latest",
+        sha256="1ead2810a0295484b5fc78a9b017f9e14152fda66ad0cecdf24e733ec2a38498",
+    )
+    version(
+        "0.4.0-gnu-12-macos-latest",
+        sha256="fc337b2b46bd664446b866d1c7a1d5cdfb1837e66d114e0d9fc6fbc1e7b47be5",
+    )
+    version(
+        "0.4.0-gnu-10-ubuntu-latest",
+        sha256="74fb14b1b2f7410cbeed0ebb854acf484989462baff46d9268f6591e83c992e7",
+    )
```
I think those should not be included here.

awvwgk replied:
```suggestion
```

awvwgk commented on repos/spack_repo/builtin/packages/mctc_lib/package.py:
```diff
@@ -23,6 +23,27 @@ class MctcLib(MesonPackage, CMakePackage):
 
     version("main", branch="main")
+
+    version("0.4.2", sha256="ce1e962c79d871d3705be590aef44f07ca296843b85e164307290f8324769406")
```
From 0.4.2 the JSON dependency changes from json-fortran to jonquil/toml-f

mtaillefumier replied:
updated. Thanks

haampie committed:
python: no recursive find in command (#386)

afzpatel committed:
yaksa: fix error when building with ROCm 6 (#337)

* yaksa: fix error when building with ROCm 6

* add comment

mtaillefumier commented:
@spackbot fix style

stephenswat committed:
geomodel: add v6.11.0 through v6.14.0 (#250)

* geomodel: add v6.11.0 through v6.14.0

This commit adds v6.11.0, v6.11.1, v6.12.0, v6.13.0, and v6.14.0 of the
geomodel package. The visualisation variant had to be disabled in the CI
to handle the newly introduced EGL dependency.

* Add gcc and clang conflicts

* gtest changes

* Testing switch

msimberg committed:
nccl: add v2.26.6 and 2.27.3 (#225)

haampie committed:
Fix style check (#410)

Commit:
Pin down the versions of multicharge and mctc-lib

mtaillefumier committed:
Merge branch 'develop' into mctc-lib

Commit:
Add the cmake build system

- simple-dftd3
- add recipy for mstore
- jonquil
- restore previous versions of toml-f

mtaillefumier commented:
@spackbot fix style

RMeli commented on repos/spack_repo/builtin/packages/jonquil/package.py:
```diff
@@ -0,0 +1,44 @@
+    version("0.2.0", sha256="68448be7f399942e15a05ed7a149cc226a8ee81a8ce66cd68a2d01d9fc86527e")
+    version("0.1.0", sha256="0c8854da047306cad357143fe56f7afe3d323d89aa7383b6614b2b587f580044")
+
+    depends_on("fortran", type="build")  # generated
```
```suggestion
    depends_on("fortran", type="build")
```

RMeli commented on repos/spack_repo/builtin/packages/mctc_lib/package.py:
```diff
@@ -23,24 +23,39 @@ class MctcLib(MesonPackage, CMakePackage):
-        return ["-Djson={0}".format("enabled" if "+json" in self.spec else "disabled")]
+        return [
+            "-Djson={0}".format("enabled" if "+json" in self.spec else "disabled"),
+            "-Dopenmp={0}".format("true" if "json" in self.spec else "false"),
```
```suggestion
            "-Dopenmp={0}".format("true" if "+openmp" in self.spec else "false"),
```
Is `true` the correct value as opposed as `enabled`/`disabled`?

RMeli commented on repos/spack_repo/builtin/packages/toml_f/package.py:
```diff
@@ -2,30 +2,46 @@
+    git = "https://github.com/toml-f/toml-f/"
 
-    maintainers("awvwgk")
+    maintainers("mtaillefumier")
```
```suggestion
    maintainers("awvwgk", "mtaillefumier")
```
? Unless @awvwgk wants to be removed...

RMeli commented on repos/spack_repo/builtin/packages/mstore/package.py:
```diff
@@ -0,0 +1,49 @@
+    variant("openmp", default=True, description="Use OpenMP parallelisation")
+
+    depends_on("c", type="build")  # generated
+    depends_on("fortran", type="build")  # generated
```
```suggestion
    depends_on("c", type="build")
    depends_on("fortran", type="build")
```

mtaillefumier committed:
Update repos/spack_repo/builtin/packages/mctc_lib/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update repos/spack_repo/builtin/packages/jonquil/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update repos/spack_repo/builtin/packages/toml_f/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

mtaillefumier committed:
Update repos/spack_repo/builtin/packages/mstore/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

Commit:
Update sirius

mtaillefumier committed:
Merge branch 'develop' into mctc-lib

Commit:
Update sirius version in cp2k recipe

mtaillefumier committed:
Merge branch 'develop' into mctc-lib

Commit:
Various fixes

RMeli commented on repos/spack_repo/builtin/packages/cp2k/package.py:
```diff
@@ -225,8 +229,8 @@ class Cp2k(MakefilePackage, CMakePackage, CudaPackage, ROCmPackage):
     depends_on("greenx", when="+greenx")
     depends_on("hdf5+hl+fortran", when="+hdf5")
     depends_on("trexio", when="+trexio")
-    depends_on("libvdwxc", when="+vdwxc")
```
The variant is still needed to explicitly set `CP2K_USE_LIBVDWXC` but the dependency is taken care of by SIRIUS, is this correct?

mtaillefumier replied:
true. Thanks for fixing this.

RMeli commented on repos/spack_repo/builtin/packages/cp2k/package.py:
```diff
@@ -337,18 +341,21 @@ class Cp2k(MakefilePackage, CMakePackage, CudaPackage, ROCmPackage):
         depends_on("sirius@7.5:", when="@2024.1:")
-        depends_on("sirius@7.6:7.7 +pugixml", when="@2024.2:")
-        depends_on("sirius@7.7: +pugixml", when="@2025.2:")
+        depends_on("sirius@7.6:+pugixml", when="@2024.2:2025.2")
+        depends_on("sirius@7.7:+pugixml", when="@2025.2:")
```
Ranges are inclusive, so these overlap. 
```suggestion
        depends_on("sirius@7.6:+pugixml", when="@2024.2:")
        depends_on("sirius@7.7:+pugixml", when="@2025.2:")
```
?

RMeli commented on repos/spack_repo/builtin/packages/dbcsr/package.py:
```diff
@@ -197,6 +204,7 @@ def cmake_args(self):
                 "gfx90a": "Mi250",
                 "gfx90a:xnack-": "Mi250",
                 "gfx90a:xnack+": "Mi250",
+                "gfx942": "Mi250",
```
```suggestion
                "gfx942": "Mi300",
```
?

mtaillefumier replied:
The dbcsr build system does not support Mi300 yet for various reasons. the `"gfx942": "Mi250",` is needed here :(

RMeli replied:
Ah, I see. In this case I think it would be better to have a conflict on `amdgpu_target=gfx942` then.

mtaillefumier replied:
I am reluctant to add a conflict for this as the only thing that this parameter controls is the file containing the optimized parameters for each GPU and the architecture during JIT time. 
The lack of optimized parameters for `Mi300` should not block users from using `dbcsr`. The only difference will be a performance "penalty" if any.  Ideally, the build system should take care of this but it is not yet present in the main repository. 
I also want to be able to compile cp2k on mi300 without having to modify the recipe locally as I am forced to do right now. 
I acknowledge though it is not the best solution.

RMeli replied:
> The lack of optimized parameters for Mi300 should not block users from using dbcsr.

Users will not be blocked to use `dbcsr`, but they will have to set `dbcsr amdgpu_target=gfx90a`. In this way, you will not magically change the meaning of setting `amdgpu_target=gfx942` once the architecture is properly supported. The user would be consciously using the parameters for another architecture (you can also add this suggestion to the conflict message).

Something like
```
cp2k +rocm amdgpu_target=gfx942 ^dbcsr +rocm amdgpu_target=gfx90a
```
should work, without the need to change the recipe.

mtaillefumier replied:
added a conflict and a message suggesting to use `gfx90a` when dbcsr is compiled for mi300. 
NB: Using parameters from older architecture should not lead to a bad behavior otherwise it means the code is buggy.

RMeli commented on repos/spack_repo/builtin/packages/dftd4/package.py:
```diff
@@ -56,6 +56,8 @@ class Dftd4(MesonPackage, CMakePackage):
         depends_on(f"mctc-lib build_system={build_system}", when=f"build_system={build_system}")
         depends_on(f"multicharge build_system={build_system}", when=f"build_system={build_system}")
 
+    depends_on("mctc-lib@0.3", when="@:3.8")
+    depends_on("multicharge@0.3", when="@:3.8")
```
It depends only on these specific versions? Or it should be a version range?

mtaillefumier replied:
The versions of the dependencies are locked unfortunately. The developers also changed the api of these libraries which lead to compilation failure for new versions of these packages.

mtaillefumier committed:
Update repos/spack_repo/builtin/packages/cp2k/package.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

Commit:
small changes for dftd4

Commit:
Change logic when dbcsr is compiled on mi300

RMeli commented on repos/spack_repo/builtin/packages/dbcsr/package.py:
```diff
@@ -108,6 +108,12 @@ class Dbcsr(CMakePackage, CudaPackage, ROCmPackage):
     dbcsr_amdgpu_targets = ("gfx906", "gfx910", "gfx90a", "gfx90a:xnack-", "gfx90a:xnack+")
     amd_msg = f"DBCSR supports these AMD gpu targets:  {', '.join(dbcsr_amdgpu_targets)}"
 
+    conflicts(
+        "+rocm",
+        when="amdgpu_target=gfx942",
+        msg="DBCSR does not support gfx942. Please use amdgpu_target=gfx90a",
+    )
+
```
Actually, I did not realize this before from your PR, but I think the conflict below already takes care of this, since `gfx942` is not added to `dbcsr_amdgpu_targets`?

Commit:
Updated the error message for unsupported amdgpu targets

RMeli commented on repos/spack_repo/builtin/packages/dbcsr/package.py:
```diff
@@ -106,13 +106,7 @@ class Dbcsr(CMakePackage, CudaPackage, ROCmPackage):
-        when="amdgpu_target=gfx942",
-        msg="DBCSR does not support gfx942. Please use amdgpu_target=gfx90a",
-    )
+    amd_msg = f"DBCSR supports these AMD gpu targets:  {', '.join(dbcsr_amdgpu_targets)}. Use amdgpu_target=gfx942 to enable gfx942 (mi300) support in DBCSR"
```
```suggestion
    amd_msg = f"DBCSR supports these AMD GPU targets:  {', '.join(dbcsr_amdgpu_targets)}. Set 'amdgpu_target' explicitly to one of the supported targets."
```
? I think this would be a bit more general than just for Mi300.

RMeli reviewed (dismissed):
LGTM. I'll let @awvwgk comment on the DFTD and DFTD-related packages.

Commit:
Add dbcsr cpu only support even if cp2k +rocm or +cuda

- Added dbcsr cpu only version when cp2k is compiled with +cuda or +rocm
- Updated cp2k version

mtaillefumier committed:
Merge branch 'spack:develop' into mctc-lib

Commit:
restore old behavior for dbcsr

mtaillefumier committed:
Merge branch 'spack:develop' into mctc-lib

bernhardkaindl commented:
@awvwgk, @RMeli wrote:
> LGTM. I'll let @awvwgk comment on the DFTD and DFTD-related packages.

Commit:
updated tblite version

mtaillefumier committed:
Merge branch 'develop' into mctc-lib

Commit:
Minor update of libvdwxc

awvwgk commented on repos/spack_repo/builtin/packages/jonquil/package.py:
```diff
@@ -0,0 +1,44 @@
+    depends_on("fortran", type="build")
+    depends_on("meson@0.57.2:", type="build", when="build_system=meson")
+
+    depends_on("toml-f")
```
Needs version 0.4 at least

Commit:
Update toml-f version for jonquil
